File: nothing/modules.py
"""Recommended Magisk module list, download, and install helpers.

Module definitions live in modules.json (skill root directory).
Edit that file to add, remove, or update modules — no Python changes needed.
"""

import json
import os
import re
from dataclasses import dataclass
from pathlib import Path
from urllib.error import URLError
from urllib.request import Request, urlopen

from .device import adb_push, confirm, run
from .exceptions import AdbError, MagiskError
from .firmware import USER_AGENT, download_file
from .models import DeviceInfo

GITHUB_API_BASE  = "https://api.github.com/repos"
# modules.json sits next to nothingctl.py (two levels up from this file)
_MODULES_JSON    = Path(__file__).parent.parent / "modules.json"


# ---------------------------------------------------------------------------
# ModuleInfo dataclass — populated at runtime from modules.json
# ---------------------------------------------------------------------------

@dataclass
class ModuleInfo:
    id:              str
    name:            str
    description:     str
    category:        str         # framework / privacy / utility / apps
    source:          str         # "github" | "ksu_store"
    install_type:    str         # "zip" | "apk"
    repo:            str | None  # "owner/repo" for GitHub
    asset_pattern:   str         # regex to match release asset filename
    requires_zygisk: bool = False
    use_prerelease:  bool = False
    notes:           str = ""


def load_modules(json_path: Path = _MODULES_JSON) -> list[ModuleInfo]:
    """Load module definitions from modules.json."""
    if not json_path.exists():
        raise FileNotFoundError(
            f"modules.json not found at {json_path}\n"
            "Expected alongside nothingctl.py in the skill directory."
        )
    with open(json_path, encoding="utf-8") as f:
        data = json.load(f)
    return [ModuleInfo(**entry) for entry in data["modules"]]


# ---------------------------------------------------------------------------
# GitHub helper — generic, any repo
# ---------------------------------------------------------------------------

def _github_headers() -> dict:
    """Build request headers, adding Authorization if GITHUB_TOKEN is set."""
    headers = {"User-Agent": USER_AGENT}
    token = os.environ.get("GITHUB_TOKEN")
    if token:
        headers["Authorization"] = f"Bearer {token}"
    return headers


def _github_latest(repo: str, use_prerelease: bool = False) -> tuple[str, list[dict]]:
    """Return (tag_name, assets) for the latest release of any GitHub repo.

    When use_prerelease=True, fetches the full releases list and returns the
    first entry (most recent), which may be a pre-release. Needed for repos
    like PlayIntegrityFix that publish only pre-releases.
    """
    if use_prerelease:
        url = f"{GITHUB_API_BASE}/{repo}/releases?per_page=1"
    else:
        url = f"{GITHUB_API_BASE}/{repo}/releases/latest"
    req = Request(url, headers=_github_headers())
    with urlopen(req, timeout=20) as resp:
        data = json.loads(resp.read())
    if use_prerelease:
        data = data[0]   # list → take first (most recent)
    return data["tag_name"], data.get("assets", [])


def _find_asset(assets: list[dict], pattern: str) -> dict | None:
    """Return first matching asset, preferring arm64 builds for Nothing phones."""
    rx      = re.compile(pattern, re.IGNORECASE)
    matches = [a for a in assets if rx.search(a["name"])]
    if not matches:
        return None
    # Nothing phones are all ARM64 — prefer arm64 when multiple builds exist
    arm64 = [a for a in matches if "arm64" in a["name"].lower()]
    return arm64[0] if arm64 else matches[0]


# ---------------------------------------------------------------------------
# Installed module detection
# ---------------------------------------------------------------------------

def get_installed_modules(serial: str) -> set[str]:
    """Return directory names under /data/adb/modules/ (requires root)."""
    r = run(["adb", "-s", serial, "shell",
             "su -c 'ls /data/adb/modules/ 2>/dev/null'"])
    if r.returncode != 0 or not r.stdout.strip():
        return set()
    return set(r.stdout.strip().splitlines())


def get_installed_versions(serial: str) -> dict[str, str]:
    """
    Read version= from every /data/adb/modules/*/module.prop in one ADB call.
    Returns {dir_name: version_string}.
    """
    r = run(["adb", "-s", serial, "shell",
             "su -c 'for d in /data/adb/modules/*/; do "
             "  name=$(basename $d); "
             "  ver=$(grep -m1 \"^version=\" $d/module.prop 2>/dev/null | cut -d= -f2); "
             "  echo \"$name|$ver\"; "
             "done'"])
    result: dict[str, str] = {}
    for line in r.stdout.strip().splitlines():
        parts = line.split("|", 1)
        if len(parts) == 2 and parts[0]:
            result[parts[0].strip()] = parts[1].strip()
    return result


def _is_installed(module: ModuleInfo, installed_dirs: set[str]) -> bool:
    """Fuzzy-match module id against installed Magisk module directory names."""
    key = module.id.lower().replace("-", "").replace("_", "")
    for d in installed_dirs:
        if key in d.lower().replace("_", "").replace("-", ""):
            return True
    return False


# ---------------------------------------------------------------------------
# Display
# ---------------------------------------------------------------------------

def _installed_version(module: ModuleInfo, installed_dirs: set[str],
                       versions: dict[str, str]) -> str | None:
    """Find installed version string by fuzzy-matching module id to dir names."""
    key = module.id.lower().replace("-", "").replace("_", "")
    for d in installed_dirs:
        if key in d.lower().replace("_", "").replace("-", ""):
            return versions.get(d)
    return None


def print_module_list(
    modules: list[ModuleInfo],
    installed_dirs: set[str],
    installed_versions: dict[str, str],
    latest_tags: dict[str, str | None],
    root_available: bool,
) -> None:
    CATEGORIES = ["framework", "privacy", "utility", "apps"]
    print()
    for cat in CATEGORIES:
        group = [m for m in modules if m.category == cat]
        if not group:
            continue
        print(f"[{cat}]")
        for m in group:
            tag         = latest_tags.get(m.id)
            latest_str  = tag or "?"

            if m.source == "ksu_store":
                status = "[manual install]"
                ver_str = ""
            elif not root_available:
                status  = "[unknown]"
                ver_str = ""
            elif _is_installed(m, installed_dirs):
                inst_ver = _installed_version(m, installed_dirs, installed_versions)
                if inst_ver and tag:
                    # Normalize for comparison: strip leading 'v'
                    iv = inst_ver.lstrip("v")
                    tv = tag.lstrip("v")
                    if iv == tv:
                        status  = "[INSTALLED]"
                        ver_str = f"  {inst_ver} — up to date"
                    else:
                        status  = "[UPDATE]"
                        ver_str = f"  {inst_ver} -> {tag}"
                else:
                    status  = "[INSTALLED]"
                    ver_str = f"  {inst_ver or '?'}"
            else:
                status  = "[not installed]"
                ver_str = f"  latest: {latest_str}"

            zygisk_str = "  [needs Zygisk]" if m.requires_zygisk else ""
            print(f"  {m.id:<24} {status:<16}{ver_str}{zygisk_str}")
            print(f"  {'':24} {m.description}")
            if m.notes:
                print(f"  {'':24} -> {m.notes}")
            print()

    print("Usage:")
    print("  List modules  :  python nothingctl.py --modules")
    print("  Install one   :  python nothingctl.py --modules --install lsposed")
    print("  Install set   :  python nothingctl.py --modules --install lsposed,shamiko,play-integrity-fix")
    print("  Install all   :  python nothingctl.py --modules --install all")


# ---------------------------------------------------------------------------
# Download + install
# ---------------------------------------------------------------------------

def _download_module(module: ModuleInfo, base_dir: Path) -> Path:
    if module.source != "github" or module.repo is None:
        raise MagiskError(
            f"'{module.id}' requires manual install.\n"
            f"  → {module.notes}"
        )
    tag, assets = _github_latest(module.repo, use_prerelease=module.use_prerelease)
    asset = _find_asset(assets, module.asset_pattern)
    if not asset:
        available = [a["name"] for a in assets]
        raise MagiskError(
            f"No matching asset for '{module.id}' in {module.repo} @ {tag}.\n"
            f"Pattern: {module.asset_pattern}\n"
            f"Available: {available}"
        )
    dest = base_dir / "modules" / module.id / tag / asset["name"]
    if dest.exists():
        print(f"  Cached  : {dest.name}")
        return dest
    mb = asset["size"] // 1024 // 1024
    print(f"  Download: {asset['name']} ({mb} MB)...")
    return download_file(asset["browser_download_url"], dest)


def _install_zip(local: Path, device: DeviceInfo) -> None:
    remote = f"/data/local/tmp/{local.name}"
    print(f"  Pushing {local.name}...")
    adb_push(local, remote, device.serial)
    print("  Installing via Magisk...")
    r = run(["adb", "-s", device.serial, "shell",
             f"su -c 'magisk --install-module {remote} && rm -f {remote}'"])
    if r.returncode != 0:
        raise MagiskError(
            f"Module install failed: {(r.stderr or r.stdout).strip()}"
        )


def _install_apk(local: Path, device: DeviceInfo, module: ModuleInfo) -> None:
    if module.notes:
        print(f"  NOTE: {module.notes}")
    print(f"  Installing APK {local.name}...")
    r = run(["adb", "-s", device.serial, "install", "-r", str(local)])
    if r.returncode != 0:
        raise AdbError(f"APK install failed: {r.stderr.strip()}")


def install_module(module: ModuleInfo, device: DeviceInfo, base_dir: Path) -> None:
    """Download and install a single module. Skips ksu_store modules with instructions."""
    if module.source == "ksu_store":
        print(f"\n[SKIP] {module.name} — manual install required.")
        print(f"       → {module.notes}")
        return

    print(f"\nInstalling {module.name}...")
    local = _download_module(module, base_dir)

    if module.install_type == "zip":
        _install_zip(local, device)
    else:
        _install_apk(local, device, module)

    print(f"[OK] {module.name} installed.")
    if module.notes:
        print(f"     → {module.notes}")


# ---------------------------------------------------------------------------
# Main action
# ---------------------------------------------------------------------------

def action_modules(
    device: DeviceInfo,
    base_dir: Path,
    install_ids: str | None,
) -> None:
    """
    List recommended modules with install status, or install specified ones.

    install_ids: None = list only | "all" = all GitHub modules | "id,id,..." = specific ones
    """
    from .backup import check_adb_root  # local import to avoid circular dependency

    modules       = load_modules()
    root_ok           = check_adb_root(device.serial)
    installed_dirs    = get_installed_modules(device.serial) if root_ok else set()
    installed_versions = get_installed_versions(device.serial) if root_ok else {}

    if not root_ok and install_ids:
        raise AdbError(
            "Root required to install Magisk modules.\n"
            "Enable in Magisk: Settings → Superuser access → Apps and ADB."
        )

    # Fetch latest release tags for display
    print("Fetching release info", end="", flush=True)
    latest_tags: dict[str, str | None] = {}
    for m in modules:
        if m.source == "github" and m.repo:
            try:
                tag, _ = _github_latest(m.repo, use_prerelease=m.use_prerelease)
                latest_tags[m.id] = tag
            except Exception:
                latest_tags[m.id] = None
        else:
            latest_tags[m.id] = None
        print(".", end="", flush=True)
    print()

    if not install_ids:
        print_module_list(modules, installed_dirs, installed_versions, latest_tags, root_ok)
        return

    # ── Install mode ──────────────────────────────────────────────────────
    if install_ids.lower() == "all":
        targets = [m for m in modules if m.source == "github"]
    else:
        ids      = {s.strip() for s in install_ids.split(",")}
        targets  = [m for m in modules if m.id in ids]
        missing  = ids - {m.id for m in targets}
        if missing:
            valid = [m.id for m in modules]
            print(f"WARNING: Unknown IDs: {', '.join(missing)}")
            print(f"Valid IDs: {', '.join(valid)}")

    if not targets:
        print("No modules to install.")
        return

    print(f"\nWill install: {', '.join(m.name for m in targets)}")
    confirm("Proceed?")

    failed = []
    for m in targets:
        try:
            install_module(m, device, base_dir)
        except (MagiskError, AdbError) as e:
            print(f"[FAIL] {m.name}: {e}")
            failed.append(m.id)

    if failed:
        print(f"\nFailed: {', '.join(failed)}")
    ok = len(targets) - len(failed)
    print(f"\n[OK] {ok}/{len(targets)} modules installed.")
    if ok:
        print("     Reboot device to activate modules.")

File: go/internal/modules/modules.go
package modules

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"time"

	"github.com/Limplom/nothingctl/internal/adb"
	"github.com/Limplom/nothingctl/internal/data"
	nterrors "github.com/Limplom/nothingctl/internal/errors"
)

const githubAPIBase = "https://api.github.com/repos"
const userAgent = "nothingctl/go"

type modulesJSON struct {
	Modules []ModuleInfo `json:"modules"`
}

type githubAsset struct {
	Name               string `json:"name"`
	Size               int64  `json:"size"`
	BrowserDownloadURL string `json:"browser_download_url"`
}

type githubRelease struct {
	TagName string        `json:"tag_name"`
	Assets  []githubAsset `json:"assets"`
}

func loadModules() ([]ModuleInfo, error) {
	var m modulesJSON
	if err := json.Unmarshal(data.ModulesJSON, &m); err != nil {
		return nil, fmt.Errorf("failed to parse modules.json: %w", err)
	}
	return m.Modules, nil
}

func githubLatest(repo string, usePrerelease bool) (string, []githubAsset, error) {
	client := &http.Client{Timeout: 20 * time.Second}
	var url string
	if usePrerelease {
		url = fmt.Sprintf("%s/%s/releases?per_page=1", githubAPIBase, repo)
	} else {
		url = fmt.Sprintf("%s/%s/releases/latest", githubAPIBase, repo)
	}

	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return "", nil, err
	}
	req.Header.Set("User-Agent", userAgent)
	if token := os.Getenv("GITHUB_TOKEN"); token != "" {
		req.Header.Set("Authorization", "Bearer "+token)
	}

	resp, err := client.Do(req)
	if err != nil {
		return "", nil, err
	}
	defer resp.Body.Close()
	body, _ := io.ReadAll(resp.Body)

	if usePrerelease {
		var releases []githubRelease
		if err := json.Unmarshal(body, &releases); err != nil || len(releases) == 0 {
			return "", nil, fmt.Errorf("failed to parse releases for %s", repo)
		}
		return releases[0].TagName, releases[0].Assets, nil
	}

	var rel githubRelease
	if err := json.Unmarshal(body, &rel); err != nil {
		return "", nil, fmt.Errorf("failed to parse release for %s", repo)
	}
	return rel.TagName, rel.Assets, nil
}

func findAsset(assets []githubAsset, pattern string) *githubAsset {
	rx := regexp.MustCompile("(?i)" + pattern)
	var matches []githubAsset
	for _, a := range assets {
		if rx.MatchString(a.Name) {
			matches = append(matches, a)
		}
	}
	if len(matches) == 0 {
		return nil
	}
	for _, a := range matches {
		if strings.Contains(strings.ToLower(a.Name), "arm64") {
			return &a
		}
	}
	return &matches[0]
}

func getInstalledModules(serial string) map[string]bool {
	stdout, _, code := adb.Run([]string{"adb", "-s", serial, "shell", "su -c 'ls /data/adb/modules/ 2>/dev/null'"})
	result := make(map[string]bool)
	if code != 0 || strings.TrimSpace(stdout) == "" {
		return result
	}
	for _, line := range strings.Split(strings.TrimSpace(stdout), "\n") {
		d := strings.TrimSpace(strings.TrimRight(line, "\r"))
		if d != "" {
			result[d] = true
		}
	}
	return result
}

func getInstalledVersions(serial string) map[string]string {
	cmd := "su -c 'for d in /data/adb/modules/*/; do name=$(basename $d); ver=$(grep -m1 \"^version=\" $d/module.prop 2>/dev/null | cut -d= -f2); echo \"$name|$ver\"; done'"
	stdout, _, _ := adb.Run([]string{"adb", "-s", serial, "shell", cmd})
	result := make(map[string]string)
	for _, line := range strings.Split(strings.TrimSpace(stdout), "\n") {
		line = strings.TrimRight(line, "\r")
		parts := strings.SplitN(line, "|", 2)
		if len(parts) == 2 && strings.TrimSpace(parts[0]) != "" {
			result[strings.TrimSpace(parts[0])] = strings.TrimSpace(parts[1])
		}
	}
	return result
}

func normalizeModKey(s string) string {
	s = strings.ToLower(s)
	s = strings.ReplaceAll(s, "-", "")
	s = strings.ReplaceAll(s, "_", "")
	return s
}

func isInstalled(m ModuleInfo, installedDirs map[string]bool) bool {
	key := normalizeModKey(m.ID)
	for d := range installedDirs {
		if strings.Contains(normalizeModKey(d), key) {
			return true
		}
	}
	return false
}

func installedVersion(m ModuleInfo, installedDirs map[string]bool, versions map[string]string) string {
	key := normalizeModKey(m.ID)
	for d := range installedDirs {
		if strings.Contains(normalizeModKey(d), key) {
			return versions[d]
		}
	}
	return ""
}

func fuzzyFindDir(moduleID string, installedDirs map[string]bool) string {
	key := normalizeModKey(moduleID)
	var sorted []string
	for d := range installedDirs {
		sorted = append(sorted, d)
	}
	for _, d := range sorted {
		if strings.Contains(normalizeModKey(d), key) {
			return d
		}
	}
	return ""
}

func printModuleList(modules []ModuleInfo, installedDirs map[string]bool, installedVersions map[string]string, latestTags map[string]string, rootAvailable bool) {
	categories := []string{"framework", "privacy", "utility", "apps"}
	fmt.Println()
	for _, cat := range categories {
		var group []ModuleInfo
		for _, m := range modules {
			if m.Category == cat {
				group = append(group, m)
			}
		}
		if len(group) == 0 {
			continue
		}
		fmt.Printf("[%s]\n", cat)
		for _, m := range group {
			tag := latestTags[m.ID]
			latestStr := tag
			if latestStr == "" {
				latestStr = "?"
			}

			var status, verStr string
			if m.Source == "ksu_store" {
				status = "[manual install]"
				verStr = ""
			} else if !rootAvailable {
				status = "[unknown]"
				verStr = ""
			} else if isInstalled(m, installedDirs) {
				instVer := installedVersion(m, installedDirs, installedVersions)
				if instVer != "" && tag != "" {
					iv := strings.TrimPrefix(instVer, "v")
					tv := strings.TrimPrefix(tag, "v")
					if iv == tv {
						status = "[INSTALLED]"
						verStr = fmt.Sprintf("  %s — up to date", instVer)
					} else {
						status = "[UPDATE]"
						verStr = fmt.Sprintf("  %s -> %s", instVer, tag)
					}
				} else {
					status = "[INSTALLED]"
					if instVer == "" {
						instVer = "?"
					}
					verStr = fmt.Sprintf("  %s", instVer)
				}
			} else {
				status = "[not installed]"
				verStr = fmt.Sprintf("  latest: %s", latestStr)
			}

			zygiskStr := ""
			if m.RequiresZygisk {
				zygiskStr = "  [needs Zygisk]"
			}
			fmt.Printf("  %-24s %-16s%s%s\n", m.ID, status, verStr, zygiskStr)
			fmt.Printf("  %-24s %s\n", "", m.Description)
			if m.Notes != "" {
				fmt.Printf("  %-24s -> %s\n", "", m.Notes)
			}
			fmt.Println()
		}
	}

	fmt.Println("Usage:")
	fmt.Println("  List modules  :  nothingctl modules")
	fmt.Println("  Install one   :  nothingctl modules --install lsposed")
	fmt.Println("  Install set   :  nothingctl modules --install lsposed,shamiko,play-integrity-fix")
	fmt.Println("  Install all   :  nothingctl modules --install all")
}

func downloadModule(m ModuleInfo, baseDir string) (string, error) {
	if m.Source != "github" || m.Repo == "" {
		return "", nterrors.MagiskError(fmt.Sprintf("'%s' requires manual install.\n  → %s", m.ID, m.Notes))
	}
	tag, assets, err := githubLatest(m.Repo, m.UsePrerelease)
	if err != nil {
		return "", nterrors.MagiskError(fmt.Sprintf("failed to fetch release for %s: %v", m.ID, err))
	}
	asset := findAsset(assets, m.AssetPattern)
	if asset == nil {
		return "", nterrors.MagiskError(fmt.Sprintf("no matching asset for '%s' in %s @ %s", m.ID, m.Repo, tag))
	}

	destDir := filepath.Join(baseDir, "modules", m.ID, tag)
	if err := os.MkdirAll(destDir, 0755); err != nil {
		return "", err
	}
	dest := filepath.Join(destDir, asset.Name)

	if _, err := os.Stat(dest); err == nil {
		fmt.Printf("  Cached  : %s\n", asset.Name)
		return dest, nil
	}

	mb := asset.Size / 1024 / 1024
	fmt.Printf("  Download: %s (%d MB)...\n", asset.Name, mb)

	client := &http.Client{Timeout: 5 * time.Minute}
	req, _ := http.NewRequest("GET", asset.BrowserDownloadURL, nil)
	req.Header.Set("User-Agent", userAgent)
	resp, err := client.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	f, err := os.Create(dest)
	if err != nil {
		return "", err
	}
	defer f.Close()
	if _, err := io.Copy(f, resp.Body); err != nil {
		return "", err
	}
	return dest, nil
}

func installZip(localPath string, serial, model string) error {
	remote := "/data/local/tmp/" + filepath.Base(localPath)
	fmt.Printf("  Pushing %s...\n", filepath.Base(localPath))
	if err := adb.AdbPush(serial, localPath, remote); err != nil {
		return err
	}
	fmt.Println("  Installing via Magisk...")
	_, stderr, code := adb.Run([]string{"adb", "-s", serial, "shell",
		fmt.Sprintf("su -c 'magisk --install-module %s && rm -f %s'", remote, remote)})
	if code != 0 {
		return nterrors.MagiskError(fmt.Sprintf("module install failed: %s", strings.TrimSpace(stderr)))
	}
	return nil
}

func installAPK(localPath, serial string, m ModuleInfo) error {
	if m.Notes != "" {
		fmt.Printf("  NOTE: %s\n", m.Notes)
	}
	fmt.Printf("  Installing APK %s...\n", filepath.Base(localPath))
	_, stderr, code := adb.Run([]string{"adb", "-s", serial, "install", "-r", localPath})
	if code != 0 {
		return nterrors.AdbError(fmt.Sprintf("APK install failed: %s", strings.TrimSpace(stderr)))
	}
	return nil
}

func installModule(m ModuleInfo, serial, baseDir string) error {
	if m.Source == "ksu_store" {
		fmt.Printf("\n[SKIP] %s — manual install required.\n", m.Name)
		fmt.Printf("       → %s\n", m.Notes)
		return nil
	}

	fmt.Printf("\nInstalling %s...\n", m.Name)
	localPath, err := downloadModule(m, baseDir)
	if err != nil {
		return err
	}

	if m.InstallType == "zip" {
		if err := installZip(localPath, serial, ""); err != nil {
			return err
		}
	} else {
		if err := installAPK(localPath, serial, m); err != nil {
			return err
		}
	}

	fmt.Printf("[OK] %s installed.\n", m.Name)
	if m.Notes != "" {
		fmt.Printf("     → %s\n", m.Notes)
	}
	return nil
}

// ActionModules lists recommended modules and/or installs them.
// installIDs: "" = list only, "all" = all github modules, "id,id,..." = specific ones.
func ActionModules(serial, baseDir string, installIDs []string) error {
	modules, err := loadModules()
	if err != nil {
		return err
	}

	rootOK := adb.CheckAdbRoot(serial)
	var installedDirs map[string]bool
	var installedVersions map[string]string
	if rootOK {
		installedDirs = getInstalledModules(serial)
		installedVersions = getInstalledVersions(serial)
	} else {
		installedDirs = make(map[string]bool)
		installedVersions = make(map[string]string)
	}

	if !rootOK && len(installIDs) > 0 {
		return nterrors.AdbError("Root required to install Magisk modules.\nEnable in Magisk: Settings → Superuser access → Apps and ADB.")
	}

	// Fetch latest release tags
	fmt.Print("Fetching release info")
	latestTags := make(map[string]string)
	for _, m := range modules {
		if m.Source == "github" && m.Repo != "" {
			tag, _, err := githubLatest(m.Repo, m.UsePrerelease)
			if err == nil {
				latestTags[m.ID] = tag
			}
		}
		fmt.Print(".")
	}
	fmt.Println()

	if len(installIDs) == 0 {
		printModuleList(modules, installedDirs, installedVersions, latestTags, rootOK)
		return nil
	}

	var targets []ModuleInfo
	if len(installIDs) == 1 && strings.ToLower(installIDs[0]) == "all" {
		for _, m := range modules {
			if m.Source == "github" {
				targets = append(targets, m)
			}
		}
	} else {
		idSet := make(map[string]bool)
		for _, id := range installIDs {
			idSet[strings.TrimSpace(id)] = true
		}
		foundIDs := make(map[string]bool)
		for _, m := range modules {
			if idSet[m.ID] {
				targets = append(targets, m)
				foundIDs[m.ID] = true
			}
		}
		for id := range idSet {
			if !foundIDs[id] {
				fmt.Printf("WARNING: Unknown ID: %s\n", id)
			}
		}
	}

	if len(targets) == 0 {
		fmt.Println("No modules to install.")
		return nil
	}

	var names []string
	for _, m := range targets {
		names = append(names, m.Name)
	}
	fmt.Printf("\nWill install: %s\n", strings.Join(names, ", "))
	if !adb.Confirm("Proceed?") {
		return nil
	}

	var failed []string
	for _, m := range targets {
		if err := installModule(m, serial, baseDir); err != nil {
			fmt.Printf("[FAIL] %s: %v\n", m.Name, err)
			failed = append(failed, m.ID)
		}
	}

	if len(failed) > 0 {
		fmt.Printf("\nFailed: %s\n", strings.Join(failed, ", "))
	}
	ok := len(targets) - len(failed)
	fmt.Printf("\n[OK] %d/%d modules installed.\n", ok, len(targets))
	if ok > 0 {
		fmt.Println("     Reboot device to activate modules.")
	}
	return nil
}

// ActionModulesStatus shows installed Magisk modules on device.
func ActionModulesStatus(serial string) error {
	if !adb.CheckAdbRoot(serial) {
		return nterrors.AdbError("Root not available via ADB shell.\nEnable in Magisk: Settings -> Superuser access -> Apps and ADB.")
	}

	installedDirs := getInstalledModules(serial)
	if len(installedDirs) == 0 {
		fmt.Println("No Magisk modules found in /data/adb/modules/.")
		return nil
	}

	checkCmd := "for d in /data/adb/modules/*/; do name=$(basename $d); if su -c 'test -f $d/disable' 2>/dev/null; then echo \"$name|disabled\"; else echo \"$name|enabled\"; fi; done"
	stdout, _, _ := adb.Run([]string{"adb", "-s", serial, "shell", checkCmd})

	states := make(map[string]string)
	for _, line := range strings.Split(strings.TrimSpace(stdout), "\n") {
		line = strings.TrimRight(line, "\r")
		parts := strings.SplitN(line, "|", 2)
		if len(parts) == 2 && strings.TrimSpace(parts[0]) != "" {
			states[strings.TrimSpace(parts[0])] = strings.TrimSpace(parts[1])
		}
	}

	if len(states) == 0 {
		for d := range installedDirs {
			states[d] = "unknown"
		}
	}

	fmt.Println("\n  Installed Magisk modules:\n")
	var sorted []string
	for d := range states {
		sorted = append(sorted, d)
	}
	// simple sort
	for i := 0; i < len(sorted)-1; i++ {
		for j := i + 1; j < len(sorted); j++ {
			if sorted[i] > sorted[j] {
				sorted[i], sorted[j] = sorted[j], sorted[i]
			}
		}
	}
	for _, moduleDir := range sorted {
		state := strings.ToUpper(states[moduleDir])
		suffix := ""
		if state == "DISABLED" {
			suffix = "  <- has disable file"
		}
		fmt.Printf("  %-40s [%s]%s\n", moduleDir, state, suffix)
	}
	fmt.Println()
	return nil
}

// ActionModulesToggle enables or disables installed Magisk modules.
func ActionModulesToggle(serial string, moduleIDs []string, enable bool) error {
	if !adb.CheckAdbRoot(serial) {
		return nterrors.AdbError("Root not available via ADB shell.\nEnable in Magisk: Settings -> Superuser access -> Apps and ADB.")
	}

	installedDirs := getInstalledModules(serial)
	if len(installedDirs) == 0 {
		return nterrors.AdbError("No Magisk modules found in /data/adb/modules/.")
	}

	for _, moduleID := range moduleIDs {
		moduleDir := fuzzyFindDir(moduleID, installedDirs)
		if moduleDir == "" {
			var dirs []string
			for d := range installedDirs {
				dirs = append(dirs, d)
			}
			return nterrors.AdbError(fmt.Sprintf("No installed module matching '%s' found.\nInstalled dirs: %s", moduleID, strings.Join(dirs, ", ")))
		}

		disableFile := fmt.Sprintf("/data/adb/modules/%s/disable", moduleDir)
		var cmd string
		if enable {
			cmd = fmt.Sprintf("su -c 'rm -f %s'", disableFile)
		} else {
			cmd = fmt.Sprintf("su -c 'touch %s'", disableFile)
		}

		_, stderr, code := adb.Run([]string{"adb", "-s", serial, "shell", cmd})
		if code != 0 {
			action := "enable"
			if !enable {
				action = "disable"
			}
			return nterrors.AdbError(fmt.Sprintf("Failed to %s module '%s': %s", action, moduleDir, strings.TrimSpace(stderr)))
		}

		if enable {
			fmt.Printf("[OK] %s enabled — reboot to apply\n", moduleID)
		} else {
			fmt.Printf("[OK] %s disabled — reboot to apply\n", moduleID)
		}
	}
	return nil
}
